File: frontend/src/api/mcp.ts
const API = (import.meta as any).env?.VITE_API_URL ?? "http://localhost:8000";
const BASE = `${API}/api/mcp`;

export interface ToolConfig {
  approval: "auto" | "confirm" | "deny";
  enabled: boolean;
}

export interface McpServer {
  name: string;
  transport: "stdio" | "sse" | "http";
  command: string | null;
  args: string[];
  url: string | null;
  env: Record<string, string>;
  tools_config: Record<string, ToolConfig>;
  enabled: boolean;
  created_at: number;
}

export interface McpServerCreate {
  name: string;
  transport: "stdio" | "sse" | "http";
  command?: string;
  args?: string[];
  url?: string;
  env?: Record<string, string>;
  tools_config?: Record<string, ToolConfig>;
  enabled?: boolean;
}

export interface McpServerPatch {
  transport?: "stdio" | "sse" | "http";
  command?: string;
  args?: string[];
  url?: string;
  env?: Record<string, string>;
  tools_config?: Record<string, ToolConfig>;
  enabled?: boolean;
}

export const listServers = (): Promise<McpServer[]> =>
  fetch(`${BASE}/`).then((r) => r.json());

export const createServer = (body: McpServerCreate): Promise<McpServer> =>
  fetch(`${BASE}/`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  }).then((r) => {
    if (!r.ok) return r.json().then((e) => Promise.reject(new Error(e.detail ?? `${r.status}`)));
    return r.json();
  });

export const patchServer = (name: string, body: McpServerPatch): Promise<McpServer> =>
  fetch(`${BASE}/${name}`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  }).then((r) => r.json());

export const deleteServer = (name: string): Promise<void> =>
  fetch(`${BASE}/${name}`, { method: "DELETE" }).then(() => undefined);

export const reloadServers = (): Promise<{ status: string; servers: string[] }> =>
  fetch(`${BASE}/reload`, { method: "POST" }).then((r) => {
    if (!r.ok) throw new Error(`${r.status}`);
    return r.json();
  });

export const getDiscoveredTools = (): Promise<
  Record<string, Array<{ name: string; description: string }>>
> => fetch(`${BASE}/tools`).then((r) => r.json());

File: backend/choreo/models/mcp.py
from typing import Literal
from pydantic import BaseModel


class ToolConfig(BaseModel):
    approval: Literal["auto", "confirm", "deny"] = "confirm"
    enabled: bool = True


class McpServerCreate(BaseModel):
    name: str
    transport: Literal["stdio", "sse", "http"] = "stdio"
    command: str | None = None
    args: list[str] = []
    url: str | None = None
    env: dict[str, str] = {}
    tools_config: dict[str, ToolConfig] = {}
    enabled: bool = True


class McpServerPatch(BaseModel):
    transport: Literal["stdio", "sse", "http"] | None = None
    command: str | None = None
    args: list[str] | None = None
    url: str | None = None
    env: dict[str, str] | None = None
    tools_config: dict[str, ToolConfig] | None = None
    enabled: bool | None = None


class McpServer(BaseModel):
    name: str
    transport: str
    command: str | None
    args: list[str]
    url: str | None
    env: dict[str, str]
    tools_config: dict[str, ToolConfig]
    enabled: bool
    created_at: int
